File: app/src/main/java/net/syncthing/lite/adapters/FoldersListAdapter.java
package net.syncthing.lite.adapters;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;

import net.syncthing.lite.R;
import net.syncthing.lite.databinding.ListviewFolderBinding;

import org.apache.commons.lang3.tuple.Pair;

import java.util.List;

import it.anyplace.sync.core.beans.FolderInfo;
import it.anyplace.sync.core.beans.FolderStats;

public class FoldersListAdapter extends ArrayAdapter<Pair<FolderInfo, FolderStats>> {

    public FoldersListAdapter(Context context, List<Pair<FolderInfo, FolderStats>> list) {
        super(context, R.layout.listview_folder, list);
    }

    @NonNull
    @Override
    public View getView(int position, View v, @NonNull ViewGroup parent) {
        ListviewFolderBinding binding;
        if (v == null) {
            binding = DataBindingUtil.inflate(LayoutInflater.from(getContext()), R.layout.listview_folder, parent, false);
        } else {
            binding = DataBindingUtil.bind(v);
        }
        FolderInfo folderInfo = getItem(position).getLeft();
        FolderStats folderStats = getItem(position).getRight();
        binding.folderName.setText(folderInfo.getLabel() + " (" + folderInfo.getFolder() + ")");
        binding.folderLastmodInfo.setText(folderStats.getLastUpdate() == null ? "last modified: unknown" : ("last modified: " + DateUtils.getRelativeDateTimeString(getContext(), folderStats.getLastUpdate().getTime(), DateUtils.MINUTE_IN_MILLIS, DateUtils.WEEK_IN_MILLIS, 0)));
        binding.folderContentInfo.setText(folderStats.describeSize() + ", " + folderStats.getFileCount() + " files, " + folderStats.getDirCount() + " dirs");
        return binding.getRoot();
    }

}

File: app/src/main/kotlin/net/syncthing/lite/adapters/FoldersListAdapter.kt
package net.syncthing.lite.adapters

import android.support.v7.widget.RecyclerView
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.ViewGroup
import net.syncthing.java.core.beans.FolderInfo
import net.syncthing.java.core.beans.FolderStats
import net.syncthing.lite.R
import net.syncthing.lite.databinding.ListviewFolderBinding
import kotlin.properties.Delegates

class FoldersListAdapter: RecyclerView.Adapter<FolderListViewHolder>() {
    var data: List<Pair<FolderInfo, FolderStats>> by Delegates.observable(listOf()) {
        _, _, _ -> notifyDataSetChanged()
    }

    var listener: FolderListAdapterListener? = null

    init {
        setHasStableIds(true)
    }

    override fun getItemCount() = data.size
    override fun getItemId(position: Int) = data[position].first.folderId.hashCode().toLong()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = FolderListViewHolder (
            ListviewFolderBinding.inflate(
                    LayoutInflater.from(parent.context), parent, false
            )
    )

    override fun onBindViewHolder(holder: FolderListViewHolder, position: Int) {
        val binding = holder.binding
        val (folderInfo, folderStats) = data[position]
        val context = holder.itemView.context

        binding.folderName = context.getString(R.string.folder_label_format, folderInfo.label, folderInfo.folderId)

        binding.lastModification = context.getString(R.string.last_modified_time,
                DateUtils.getRelativeDateTimeString(context, folderStats.lastUpdate.time, DateUtils.MINUTE_IN_MILLIS, DateUtils.WEEK_IN_MILLIS, 0))

        binding.info = context.getString(R.string.folder_content_info, folderStats.describeSize(), folderStats.fileCount, folderStats.dirCount)

        binding.root.setOnClickListener {
            listener?.onFolderClicked(folderInfo, folderStats)
        }
    }
}

class FolderListViewHolder(val binding: ListviewFolderBinding): RecyclerView.ViewHolder(binding.root)

interface FolderListAdapterListener {
    fun onFolderClicked(folderInfo: FolderInfo, folderStats: FolderStats)
}